include/revng/Support/GraphAlgorithms.h

```diff
@@ -406,138 +312,117 @@ getBackedgesBlackList(GraphT Block,
+    ForwardDFSNodes.insert(Block);
   }
 
-  auto Targets = State.getReachables();
-  // Add in a fixed point fashion the additional nodes
-  SmallSetVector OldTargets;
-  do {
-    // At each iteration obtain a copy of the old set, so that we are able to
-    // exit from the loop as soon no change is made to the `Targets` set
+  // The `nodesBetween` primitive can be invoked with a null `Target` in order
+  // to perform a search that simply collects all the nodes reachable from the
+  // `Source` node
+  if (Target == nullptr) {
+    return ForwardDFSNodes;
+  }
 
-    OldTargets = Targets;
+  // Perform the backward DFS visit, stopping at the `Source` node
+  ExtType BackwardExtSet;
+  BackwardExtSet.insert(Source);
 
-    // Temporary storage for the nodes to add at each iteration, to avoid
-    // invalidation on the `Targets` set
-    SmallSetVector NodesToAdd;
+  using BGT = llvm::GraphTraits<llvm::Inverse<GraphDirection>>;
+  using bdf_iterator = llvm::df_iterator<GraphT, ExtType, true, BGT>;
+  auto BackwardBegin = bdf_iterator::begin(Target, BackwardExtSet);
+  auto BackwardEnd = bdf_iterator::end(Target, BackwardExtSet);
 
-    for (NodeRef Block : Targets) {
-      SmallSetVector &AdditionalSet = State.getAdditional(Block);
-      NodesToAdd.insert(AdditionalSet.begin(), AdditionalSet.end());
-    }
+  for (NodeRef Block : llvm::make_range(BackwardBegin, BackwardEnd)) {
+    BackwardDFSNodes.insert(Block);
+  }
 
-    // Add all the additional nodes found in this step
-    Targets.insert(NodesToAdd.begin(), NodesToAdd.end());
-    NodesToAdd.clear();
+  // Perform the set intersection between the forward and backward set nodes
+  SmallSetVector Result = llvm::set_intersection(ForwardDFSNodes,
+                                                 BackwardDFSNodes);
+
+  // Corner case handling
+  if (not Result.empty()) {
+
+    // We found some nodes in the intersection of the two DFSs, therefore we
+    // actually have some reachables nodes, and we add both the `Source` and the
+    // `Target` in the `Result` set (they are not automatically included because
+    // the are part of the respective ext sets, and therefore not present in the
+    // intersection)
+    Result.insert(Target);
+    Result.insert(Source);
+  } else if (hasSuccessor<GraphT, FGT>(Source, Target)) {
+
+    // If the `Target` is a direct successor of `Source`, we may have that the
+    // resulting intersection is empty, if the only path connecting them is the
+    // trivial one, but the reachable nodes are actually `Source` and `Target`
+    revng_assert(Result.empty());
+    Result.insert(Target);
+    Result.insert(Source);
+  } else if (Source == Target) {
+
+    // If `Source` is equal to `Target`, due to fact that `Target` (which is `==
+    // Source`) is added to the ext set, (and vice-versa), we will not find any
+    // candidate note in our exploration, therefore we need to manually add it
+    // to the `Result` set
+    Result.insert(Source);
+  }
```

Comment:
Seems very nice!

Reply:
Eh, unfortunately due to the fact that the intersection will not contain the `Source` and the `Target` (they are mutually excluded by the fact they are inserted in the ext set, therefore not in the intersection), we need to manually add them afterwards (even in the corner case of when `Source == Target`.